conn_test.go

```diff
@@ -775,42 +849,54 @@ func TestResumption(t *testing.T) {
 	client2.Read(nil)
 	<-done
 
-	assertDeepEquals(t, client2.state.Params, server2.state.Params)
-	assertCipherSuiteParamsEquals(t, client2.state.cryptoParams, server2.state.cryptoParams)
-	assertByteEquals(t, client2.state.resumptionSecret, server2.state.resumptionSecret)
-	assertByteEquals(t, client2.state.clientTrafficSecret, server2.state.clientTrafficSecret)
-	assertByteEquals(t, client2.state.serverTrafficSecret, server2.state.serverTrafficSecret)
-	assert(t, client2.state.Params.UsingPSK, "Session did not use the provided PSK")
+	checkConsistency(t, client2, server2)
+	assertTrue(t, client2.state.Params.UsingPSK, "Session did not use the provided PSK")
 }
 
-func Test0xRTT(t *testing.T) {
-	conf := pskConfig
-	cConn, sConn := pipe()
-
-	client := Client(cConn, conf)
-	client.EarlyData = []byte("hello 0xRTT world!")
-
-	server := Server(sConn, conf)
-
-	done := make(chan bool)
-	go func(t *testing.T) {
-		alert := server.Handshake()
-		assertEquals(t, alert, AlertNoAlert)
-		done <- true
-	}(t)
+func test0xRTT(t *testing.T, name string, p testInstanceState) {
+	conf := *pskConfig
+	conf.NonBlocking = true
 
-	alert := client.Handshake()
-	assertEquals(t, alert, AlertNoAlert)
+	if p.get("dtls") == "true" {
+		conf.UseDTLS = true
+	}
 
-	<-done
+	cConn, sConn := pipe()
+	cbConn := newBufferedConn(cConn)
+	cbConn.SetAutoflush()
+	sbConn := newBufferedConn(sConn)
+	sbConn.SetAutoflush()
+
+	client := Client(cbConn, &conf)
+	server := Server(sbConn, &conf)
+
+	client.Handshake() // This sends CH
+	zdata := []byte("ABC")
+	n, err := client.Write(zdata) // This should succeeed
+	assertNotError(t, err, "Client was able to write")
+	assertEquals(t, n, len(zdata))
+	hsUntilBlocked(t, server, sbConn) // Read CH and early data.
+	tmp := make([]byte, 10)
+	n, err = server.Read(tmp)
+	assertNotError(t, err, "Error reading early data")
+	tmp = tmp[:n]
+	assertByteEquals(t, zdata, tmp)
+	hsRunHandshakeOneThread(t, client, server)
+
+	assertTrue(t, client.state.Params.UsingEarlyData, "Session did not negotiate early data")
+	n, err = server.Read(tmp)
+	assertEquals(t, AlertWouldBlock, err)
+	assertEquals(t, 0, n)
+}
 
-	assertDeepEquals(t, client.state.Params, server.state.Params)
-	assertCipherSuiteParamsEquals(t, client.state.cryptoParams, server.state.cryptoParams)
-	assertByteEquals(t, client.state.resumptionSecret, server.state.resumptionSecret)
-	assertByteEquals(t, client.state.clientTrafficSecret, server.state.clientTrafficSecret)
-	assertByteEquals(t, client.state.serverTrafficSecret, server.state.serverTrafficSecret)
-	assert(t, client.state.Params.UsingEarlyData, "Session did not negotiate early data")
-	assertByteEquals(t, client.EarlyData, server.EarlyData)
+func Test0xRTT(t *testing.T) {
+	params := []testParameter{
+		testParameter{
+			"dtls",
+			[]string{"true", "false"},
+		},
+	}
```

Comment:
If you change from `[]testParameter` to `type testParams map[string][]string`, then you can write things like this much more compactly:

```
testParams{
  "dtls": {"true", "false"},
}
```